src/Host/Broker/Impl/Sessions/SessionManager.cs

```diff
@@ -59,6 +60,41 @@ ILogger<Process> hostOutputLogger
             }
         }
 
+        public void CloseAndBlockSessionsCreationForUser(IIdentity user) {
+            lock (_sessions) {
+                lock (_blockedUsers) {
+                    var userSessions = GetOrCreateSessionList(user);
+                    var sessions = userSessions.ToArray();
+                    foreach (var session in sessions) {
+                        try {
+                            userSessions.Remove(session);
+                            Task.Run(() => session.KillHost()).DoNotWait();
+                        } catch (Exception) { }
+
+                        session.State = SessionState.Terminated;
+                    }
+
+                    if (!_blockedUsers.Contains(user.Name)) {
+                        _blockedUsers.Add(user.Name);
+                    }
+                }
+            }
+        }
+
+        public void UnblockSessionCreationForUser(IIdentity user) {
+            lock (_blockedUsers) {
+                if (_blockedUsers.Contains(user.Name)) {
+                    _blockedUsers.Remove(user.Name);
+                }
+            }
+        }
+
+        public bool IsSessionCreationBlockedForUser(IIdentity user) {
```

Comment:
There's no way to use this API safely - by the time it returns, it has released the lock, and the return value might already be invalid (if session creation was blocked in the meantime).

I don't think this needs to be exposed at all, actually. Just lock and make the check directly in `CreateSession`, and throw if it was denied. But the crucial point is that when session is created, it should take the lock, check, and if the check passed, then hold the lock until it finishes session creation (so that any code trying to come and remove profile would have to wait until that completes).

This will also be easier to do if `_sessions` and `_blockedUsers` share a single lock, since `CreateSession` already acquires the former for most of its body.